### ISSUE TITLE
Highlighting of special code comment types like TODO, INFO,  etc.

### ISSUE THREAD

**Issue by adiepenbrock** [zed-industries/zed#11895]
### Check for existing issues

- [X] Completed

### Describe the feature

One feature that I find really useful in other editors, e.g., Neovim with the plugin "folke/todo-comments," is the ability to highlight specific types of comments like `// TODO` `// INFO` and others. This feature enhances productivity by making it easier to locate and manage important comments within the codebase when scrolling / jumping through the code.

Because there may be users who don't want this kind of highlights in the code, it should be configurable by the user (activate / deactivate, configure highlighting color or deriving it from the used theme, etc.)

There is already a feature request (#4808) to introduce a separate tree-structure overview of these comments. However, I think this should also be available in the code itself.

### If applicable, add mockups / screenshots to help present your vision of the feature

<img width="545" alt="image" src="https://github.com/zed-industries/zed/assets/2057587/4ec6317f-b6e9-408b-93de-4c2c8ba551cd">

**Comment by francesco-gaglione** [comment]
I'm considering to work on this. Some instructions on how to do it? Colours for example

**Comment by diogox** [comment]
I don't know about the color requirements, but any implementation would probably be a good start. I use this every day in other editors and find it very useful :)

**Comment by francesco-gaglione** [comment]
I readed somewhere that this should be done zed extensions and not in the core codebase. It is still true or we want to consider to implement it in the core?

**Comment by kofron** [comment]
@francesco-gaglione did you wind up taking a stab at this?

**Comment by francesco-gaglione** [comment]
@kofron I need first the confirmation from mainteiners to do this into zed core and not in the plugins

**Comment by everdrone** [comment]
There is this open issue: https://github.com/zed-industries/extensions/issues/523 where the direction was to implement it as an extension.

**Comment by aretrace** [comment]
My hope is that a straightforward comment tag system could just be integrated into zed itself.

**Comment by francesco-gaglione** [comment]
@berkus it is a lot of time I don't contribute here... but it is very usefull so I'll take a look.... some hint on where to start?

**Comment by berkus** [comment]
No idea, i am not super familiar with zed's internals. I would look through syntax highlighting and editing plumbing - it's probably there at the intersection of those parts.

**Comment by phistep** [comment]
I also use `NOTE` and this package https://github.com/folke/todo-comments.nvim includes

```
    TODO = { icon = " ", color = "info" },
    HACK = { icon = " ", color = "warning" },
    WARN = { icon = " ", color = "warning", alt = { "WARNING", "XXX" } },
    PERF = { icon = " ", alt = { "OPTIM", "PERFORMANCE", "OPTIMIZE" } },
    NOTE = { icon = " ", color = "hint", alt = { "INFO" } },
    TEST = { icon = "⏲ ", color = "test", alt = { "TESTING", "PASSED", "FAILED" } },
```

**Comment by iairu** [comment]
> @kofron I need first the confirmation from mainteiners to do this into zed core and not in the plugins

@kofron @berkus see https://github.com/zed-industries/zed/pull/9082#issuecomment-1997957426

tl;dr: someone tried to merge a working implementation but got denied, devs want this as an extension and not in core

**Comment by Trikzon** [comment]
Not sure if this would need another issue, but it would also be nice if Zed would have a panel for listing all TODO comments in a project to find them easier.

**Comment by berkus** [comment]
> Not sure if this would need another issue, but it would also be nice if Zed would have a panel for listing all TODO comments in a project to find them easier.

This seems to be related https://github.com/zed-industries/zed/issues/10546 but it's all unified under https://github.com/zed-industries/extensions/issues/523

**Comment by sammypanda** [comment]
it would be neat if there was an option to configure zed for them to show in project diagnostics too.

**Comment by sammypanda** [comment]
> Not sure if this would need another issue, but it would also be nice if Zed would have a panel for listing all TODO comments in a project to find them easier.

it would be neat if there was the option to configure zed for them to show in project diagnostics; i think it fits nicely.

**Comment by itpropro** [comment]
Any updates from the zed team about this? Either an implementation for native support for TODO/FIXME/etc. comments or an extension for this (although I think this should be part of core as it is to wide spread) would increase the developer experience a lot. I don't think I have a single project, independent of the language, that doesn't have `TODO` comments in it.

**Comment by thangckt** [comment]
It would better to have an extension like [comment anchor](https://github.com/StarlaneStudios/vscode-comment-anchors)

**Comment by boozook** [comment]
> It would better to have an extension like [comment anchor](https://github.com/StarlaneStudios/vscode-comment-anchors)

Totally agree. And I also lack the functionality to put bookmarks (anchors) in any random place, jump onto it, list it all.

**Comment by thangckt** [comment]
> bookmarks (anchors) in any random place

That is the most of my favorite. No matter what language you are working on, just put an "anchor" and then list them in outline.

**Comment by jshrssll** [comment]
Is this feature likely to be merged?

**Comment by DZinin-Bro** [comment]
Or like https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree

**Comment by l-mb** [comment]
I'm considering zed, and I'm used to this particular feature in my previous editor (neovim).

It would be really a boon to have this (the anchor extension would also work great, as would tags). The absence makes larger edits (where I drop notes to myself as I go through the code while jumping around) less awesome.

**Comment by umbrageodotus** [comment]
+1

**Comment by diogox** [comment]
This is now [done](https://github.com/dargor/zed-python-requirements/issues/3) btw

**Comment by hyunbinseo** [comment]
tl:dr; Update Zed and install the [`Comment` extension](https://github.com/thedadams/zed-comment) from the store.

> Search "comment" in extension page, and install it. See the Compatibility section for details on what is needed from _each_ language for proper highlighting.

Works on v0.210.4:

```
Zed: v0.210.4 (Zed) 
OS: Windows 10.0.26200
Memory: 31.5 GiB
Architecture: x86_64
GPU: Intel(R) Arc(TM) Graphics || Intel Corporation || 32.0.101.6913
```

**Comment by skewty** [comment]
I am not accustomed to the required ':' suffix to trigger highlighting. Is that normal / typical?

**Comment by acgxv** [comment]
> I am not accustomed to the required ':' suffix to trigger highlighting. Is that normal / typical?

+1, and I prefer the style in this issue description. I only want the keywords’ background to be highlighted.

**Comment by everdrone** [comment]
> > I am not accustomed to the required ':' suffix to trigger highlighting. Is that normal / typical?
> 
> +1, and I prefer the style in this issue description. I only want the keywords’ background to be highlighted.

+1, I am used to the keyword only.
Most TODOs and FIXMEs span multiple lines. currently only the first line is highlighted (except the `:`).
Also, it would be super helpful to show an indicator in the gutter. Vscode extensions add customizable icons in the gutter at the line where the comment keyword appears.

**Comment by skewty** [comment]
> Most TODOs and FIXMEs span multiple lines

I believe one indicates the next line as part of the previous with a ' ' space prefix before the first letter.  Compare to the line + column where 'T' in "TODO" was written.

Example:
```python
# TODO make this part of service configuration
#  and prevent values < 5
MAX_FOO = 10
```
or
```python
#TODO make this part of service configuration
# and prevent values < 5
MAX_FOO = 10
```

Both lines, the whole comment, can be highlighted as the TODO using this trick. This is how it works in JetBrains IDEs.

**Comment by everdrone** [comment]
@skewty this is not supported by the comments extension currently

<img width="430" height="138" alt="Image" src="https://github.com/user-attachments/assets/7b85706b-2818-4b9f-9ec2-04863bd7066c" />

at least not in rust

**Comment by williamdes** [comment]
> I believe one indicates the next line as part of the previous

I kind of agree, but this rule is difficult to know if everyone agrees.
See: https://github.com/search?q=repo%3Aphpmyadmin%2Fphpmyadmin+todo&type=code
We have some lists of items and sentences with `,` at the end

**Comment by zelenenka** [comment]
closing this since there's an extension now, leave a comment if there's something else to do.

**Comment by blueray453** [comment]
@zelenenka can you please give a link to the extension.

**Comment by williamdes** [comment]
See thread https://github.com/zed-industries/extensions/issues/523#issuecomment-3488340683
